dali/python/nvidia/dali/pipeline.py

```diff
@@ -1154,20 +1154,25 @@ def _run_input_callbacks(self):
         if self._input_callbacks is None:
             return
 
+        batches = []   # data from external source callbacks is gathered here
         stop_iter = False
         for i, group in enumerate(self._parallel_input_callbacks):
             try:
-                group.schedule_and_feed(self, self._py_pool, i, self._max_batch_size)
+                batches.append(group.schedule_and_receive(self, self._py_pool, i, self._max_batch_size))
             except StopIteration:
                 stop_iter = True
         for group in self._seq_input_callbacks:
             try:
-                group.call_and_feed(self, self._max_batch_size)
+                batches.append(group.get_batch(self, self._max_batch_size))
             except StopIteration:
                 stop_iter = True
         if stop_iter:
             raise StopIteration()
 
+        # we only fill external source queues when we know that all callbacks succeeded
```

Comment:
What happens in case of failure? Are we loosing the data from all but failed callbacks?

Reply:
The whole iteration is discarded - that's the point of this PR.

Reply:
Maybe we should add this to our docs. Imagine that multiple ESes consume data from the network, I wonder if the user wants to lose data from them if any raises StopIteration. Maybe when we restart we should not run ES that has data stored already?